### PR TITLE
Pouring (Core Filling Functionality)

### LINKED ISSUE #2: Add tests and determine testing guideline/standard
Basic unit tests and integration tests should be added for the various functions in didymus, in pebble.py, mc_reader.py, and the in-development pack.py and core.py.

(this issue is also for general discussion of tests/a testing suite, including what to test and how to test it).

This issue can be closed when tests are added for what is currently in didymus, and some sort of standard/guideline for testing going forward has been determined.

### LINKED ISSUE #4: Update to pytest
Use pytest rather than nose. This issue can be closed with a PR that converts nose tests to pytest (or some other currently maintained test framework).

### LINKED ISSUE #10: Get LAMMPS input and results in didymus repo
This issue can be closed when a dedicated place for LAMMPS models has been made, and the current LAMMPS testing model has been put there, alongside its results.

### PR DESCRIPTION
**Pull request by ZoeRichter**
## Summary of changes
This PR adds the "pour" functionality to ghastly, i.e., the ability to fill a core with pebbles.  It uses a combination of OpenMC to pack cylindrical regions, and LAMMPS to both allow the pebbles to fall under gravity, and add any remaining pebbles needed to fill the core to the desired packing fraction.

It does this using a series of jinja2 templates, which automatically create a LAMMPS input file based on the parameters in the input file, which is a JSON file.  There are example input files for ghastly both in the tests and examples directories.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Associated Issues and PRs
- closes #2 
- closes #4
- closes #10

## Checklist for Reviewers

Reviewers should use [this link](https://arfc.github.io/manual/guides/pull_requests) to get to the
Review Checklist before they begin their review.

### CODE REVIEW

File: examples/pre-made-lammps/README.md
Comment (LukeSeifert) [inline examples/pre-made-lammps/README.md:9]: What are type 1 and type 2 pebbles?

File: examples/pre-made-lammps/pour-220000-peb.txt
Comment (LukeSeifert) [inline examples/pre-made-lammps/pour-220000-peb.txt:163]: Is there a way to change this into a function or do a loop?

File: ghastly/__about__.py
Comment (LukeSeifert) [inline ghastly/__about__.py:3]: No version 0.2.0?

File: ghastly/core.py
Comment (LukeSeifert) [inline ghastly/core.py:11]: `open_bottom` not in docstring parameters

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:28]: Does this mean users should supply a `rough_pf` that is greater, or that they should be careful if they do so?

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:50]:
```suggestion
    if pebbles_left < 0:
```
Since this should always be an integer

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:118]:
```suggestion
            raise TypeError("down_flow should be True or False")
```

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:177]: No return block in docstring

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:193]:
```suggestion
def find_box_bounds(sim_block, pour=False):
```

File: ghastly/read_input.py
Comment (LukeSeifert) [inline ghastly/read_input.py:42]:
```suggestion
        -------
```

File: ghastly/read_input.py
Comment (LukeSeifert) [inline ghastly/read_input.py:65]:
```suggestion
        -------
```

File: ghastly/read_input.py
Comment (LukeSeifert) [inline ghastly/read_input.py:116]:
```suggestion
        -------
```

File: ghastly/ghastly.py
Comment (LukeSeifert) [inline ghastly/ghastly.py:210]: No return block in docstring

File: ghastly/read_input.py
Comment (LukeSeifert) [inline ghastly/read_input.py:63]: Missing parameters block in docstring

File: ghastly/read_input.py
Comment (LukeSeifert) [inline ghastly/read_input.py:134]:
```suggestion
            raise TypeError('''down_flow should be True for downward flow
                            and False for upward flow''')
```

File: tests/test_ghastly.py
Comment (LukeSeifert) [inline tests/test_ghastly.py:14]:
```suggestion
                                   rough_pf=0.2)
```

**Review by LukeSeifert — COMMENTED** [review]
Overall, looks good! I had a couple comments, mostly on docstrings and some whitespace for pep8 formatting. I think if you run autopep8, it'll probably automatically catch a couple of my comments.

**Review by nsryan2 — COMMENTED** [review]
Hi Zoe! Lots of good work here, I have some minor consistency comments and questions.

Before I approve, I'd love to see Pep8 run on this PR. I noticed several instances of missing white space around math operations and other things that it won't like.

File: examples/pre-made-lammps/README.md
Comment (ZoeRichter) [reply examples/pre-made-lammps/README.md:1]: I am using a different version number for the dataset of pebble centroids than I am for ghastly, the python package.  This is part because I also want these datasets on Zenodo, and want to have a unique version number for that.

File: ghastly/core.py
Comment (ZoeRichter) [reply ghastly/core.py:28]: The original way was a bug, it was always meant to be the height (or length?) of that section of the core.  The original way only works for something centered on the origin.  I'm not against including derived attributes like height or volume in the docstring though.  Do you think I should make a separate section for them?

File: ghastly/ghastly.py
Comment (ZoeRichter) [reply ghastly/ghastly.py:1]: Given the number of output files that LAMMPS will produce, I think the best-case is for a user to use separate directories for each individual simulation, and run ghastly within that directory.  Maybe I could create an optional suffix for the generated file names?  I'm not sure if that would be clearer than just keeping things organized in directories, though.

I can fix the functions if autopep 8 does not :)

File: ghastly/ghastly.py
Comment (ZoeRichter) [reply ghastly/ghastly.py:39]: I think you are right!   Definitely overthought that one.

File: examples/pre-made-lammps/README.md
Comment (ZoeRichter) [reply examples/pre-made-lammps/README.md:9]: Generally speaking, if I refer to a 'type', I'm referring to the particle type in LAMMPS.  You use types to distinguish particles with different parameters, like size or density.  In the case of this specific dataset, the type was used to make them different colors.

File: examples/pre-made-lammps/pour-220000-peb.txt
Comment (ZoeRichter) [reply examples/pre-made-lammps/pour-220000-peb.txt:163]: Not that I've been able to find, unfortunately.  LAMMPS doesn't have loops.  For the future, I could put a block like this into a separate file, and then insert it when I need to use it (which is how I generate files with templates - it's a like lot inserting files with LaTeX).  Since this was the input file that created the associated pebble dataset, though, I don't want to change this specific input file.

File: ghastly/__about__.py
Comment (ZoeRichter) [reply ghastly/__about__.py:3]: I could have sworn last version was 0.2.0, but it looks like I was wrong.  I can set it back to 0.2.0

File: ghastly/ghastly.py
Comment (ZoeRichter) [reply ghastly/ghastly.py:30]: That's a good point.  Do you think I should still call that section of the docstring 'Returns', or do you think something like 'Generated Files' would be clearer?

**Comment by ZoeRichter** [comment]
I think I saved it - sorry folks!

File: ghastly/ghastly.py
Comment (ZoeRichter) [reply ghastly/ghastly.py:28]: Both you and Luke asked this question, so I definitely need to make this clearer - but yes, if you add a much higher rough_pf for doing OpenMC packing than you give for the target pf (like, if you only wanted 30% of the core to have pebbles, for some reason, but you supply a rough_pf of 60%) you could "overfill" the core.

In the future this will change a bit, because I'd like to adjust it to auto-fill the "discharge" region to 60% no matter what.  Then, the packing fractions (both the rough pf and the 'target' pf in the input file) would only be referring the the main body of the core (however, you can still have this "overfilling" issue).  It's also possible I remove rough_pf later, and have it always match the "target", but for now while I'm developing things I think it's potentially useful to have them separate.

File: ghastly/core.py
Comment (nsryan2) [inline ghastly/core.py:83]:
```suggestion
        raise NotImplementedError("Annular core regions are not implemented.")
```

File: ghastly/core.py
Comment (nsryan2) [inline ghastly/core.py:141]:
```suggestion
        raise NotImplementedError("Annular core regions are not implemented.")
```

File: ghastly/ghastly.py
Comment (nsryan2) [reply ghastly/ghastly.py:1]: That's a good point about the number of output files, I maybe over name my files.

File: ghastly/simulation.py
Comment (nsryan2) [inline ghastly/simulation.py:48]:
```suggestion
        self.pebble_volume = (4/3) * np.pi * (r_pebble**3)
```
conflict with pep8

File: ghastly/ghastly.py
Comment (nsryan2) [reply ghastly/ghastly.py:54]: It's also not stylistically consistent with the rest of your PR, which doesn't contain print statements.

File: ghastly/ghastly.py
Comment (ZoeRichter) [reply ghastly/ghastly.py]: 'rough_' is what I was using to denote they are different - the 'target' packing fraction is just called 'pf' in the input.

File: examples/pre-made-lammps/README.md
Comment (katyhuff) [inline examples/pre-made-lammps/README.md:14]: dumpfile is two words earlier. be consistent.

File: examples/pre-made-lammps/README.md
Comment (katyhuff) [inline examples/pre-made-lammps/README.md:16]: Please include the LAMMPS command you ran, with all option flags, in this readme, so that someone could reproduce your output from your input. Alternatively, write a little shell/python script.

File: ghastly/ghastly.py
Comment (katyhuff) [reply ghastly/ghastly.py:54]: Agree. Though, I think this is optional for now. Consider creating an issue about this to do as future work. This is the sort of thing that shouldn't hold up a PR merge, but could be noted as future work (even something an undergrad could do on their own).

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:70]: The next 12 lines should be it's own function. Call it "generate_variable_block" or something.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:82]: The next section should be its own function. Call it. "generate_regions_block"

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:64]: This should probably be called coords_to_lammps_dumpfile or similiar. It would be ok to have it be "write_dumpfile" though it may be confusing as that implies it would be a ghastly dumpfile. Maybe "write_lammps_dumpfile" is sufficient.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:65]: Consider making a point class so that all x,y,z triplets are checkable and passable as single objects.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:118]: Here again, the following section should be its own function.

File: ghastly/ghastly.py
Comment (katyhuff) [reply ghastly/ghastly.py:65]: this can be a future issue. Consider the best way to represent geometries.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:237]: Create a geometry handling issue to simplify this kind of logic.

Ok for now as long as it's tested. 
Consider loading an external library (sympy? shapely? pyeuclid?)

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:256]: Maybe make these percentages input variables.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:257]: Maybe make this 20% an input variable.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:259]: move out of case switch.

File: ghastly/ghastly.py
Comment (katyhuff) [inline ghastly/ghastly.py:303]: I recommend "pebble_coords"

**Review by katyhuff — APPROVED** [review]
Minor changes suggested, most can be handled by creating issues to be dealt with in the future. In fact, I'm going to list this as "approve" so that you can merge as soon you feel you have completed the essential tasks (line lengths, etc.) and created the issues recommended. I see no fatal flaws in this PR, just some things you might consider changing.

File: ghastly/ghastly.py
Comment (nsryan2) [inline ghastly/ghastly.py:185]:
```suggestion
        cylindrical region matching the core element passed to pack_cyl.
```

**Comment by ZoeRichter** [comment]
I believe I have either made issues for or fixed issues in new comments!  (Thanks all)

**Comment by ZoeRichter** [comment]
Oops - I missed a few comments because I missed some hidden comments when I when through.  Let me make a commit for splitting up functions into smaller sections, and THEN I think I actually covered everything.

**Comment by nsryan2** [comment]
@katyhuff do you have any insight on why Pep8speaks might not have checked this PR?

File: tests/test_ghastly.py
Comment (nsryan2) [inline tests/test_ghastly.py:37]:
```suggestion
    assert z_b["low"] == pytest.approx(0.55 - 1.05*0.65)
```
Not necessary, but stylistically consistent

File: tests/test_simulation.py
Comment (nsryan2) [inline tests/test_simulation.py:20]:
```suggestion
    assert test_sim.seed >= 1000000 and test_sim.seed <= 100000000
```
Should this be `and` and not `or`?

Comment (ZoeRichter) [reply tests/test_simulation.py:20]: oh, you're right!

**Pull request merged**
